### DISCUSSION TITLE
esp32/pwm: Enhance of PWM feature to support ligth sleep mode and selecting clock source of timer.

### DISCUSSION THREAD

**Discussion by yoann-darche** [micropython/micropython#16033] [ESP32]
Does it possible to add the possibnility to change the clock source of the timer used in the LEDC configuration (for PWM python class). 
Maybe, we can add an api entry to set up the timer clock source, then add the possibility to choose the timer source in the PWM constructor.
I would like to use the LEDC_USE_RC_FAST_CLK  in order to maintain the PWM signal when the ESP32-S3 enter in LightSleep mode. In this ESP32-S3's mode only the the RC_FAST_CLK still running.

sample code that demonstarte the PWM signal lost in ligth  sleep (MicroPython v1.23.0 on 2024-06-02; Generic ESP32S3 module with ESP32S3):

````py
import esp32
import machine
from machine import Pin, Timer, PWM, lightsleep
from neopixel import NeoPixel
import time


# PWM
PWM_1 = const(4)
PWM_2 = const(13)

O_PWM1 = None
O_PWM2 = None


LED = None


# Internal Led
INTERNAL_LED = const(47)


def init_hardware():
    
    global O_PWM1, O_PWM2, LED
    
    O_PWM1 = PWM(Pin(PWM_1,  Pin.OUT), freq=25000, duty_u16=65535)
    O_PWM2 = PWM(Pin(PWM_2,  Pin.OUT), freq=12500, duty_u16=65535, mode=1)
    LED = NeoPixel(Pin(INTERNAL_LED, Pin.OUT), 1)

def Led(status=False, cGRB=(0,125,0)):
        
        global LED
        
        if LED is None:
            return
        
        if status:
            LED[0] = cGRB
        else:
            LED[0] = (0,0,0)
            
        LED.write()  

# Main

def Main():
    
    global O_PWM1, O_PWM2
    
    init_hardware()
    
    c = machine.reset_cause()
    if c == machine.PWRON_RESET:                
        Led(True,(64,0,0)) # Green
    elif c == machine.HARD_RESET:
        print("HARD_RESET")        
        Led(True,(64,128,0)) # Yellow
    elif c == machine.SOFT_RESET:
        print("SOFT_RESET")        
        Led(True,(0,128,64)) # Pink
    elif c == machine.WDT_RESET:
        print("WDT_RESET")        
    elif c == machine.DEEPSLEEP_RESET:
        print("DEEPSLEEP_RESET")        
    else:
        print("UNKNOW RESET")        
        hwd.Led(True,(0,0,64)) # Blue
        

    V = 50

    while True:
        
        if V == 50:
            V = 25
        else:
            V = 50
            
        # Set Fan PWM:
        
        duty_pct = (100-V) * 65535 // 100    
    
        O_PWM1.duty_u16(duty_pct)
        O_PWM2.duty_u16(duty_pct)
        print('OPWM1',O_PWM1)
        print('OPWM2',O_PWM2)
        
        
        time.sleep(10)
        
        print("[i] Activation du someil légé (20s) ! zZz ")
        Led(True, (0,64,0))                 
        lightsleep(20*1000)
        print("[i] WakeUp !")
        Led(True, (32,0,0))  
        

# Global Start
Main()
````

**Comment by sosi-deadeye** [comment]
It is totally mind-blowing, that an MCU is capable, in light sleep, to control a PWM output. Is it possible to change the clock src with manipulating the memory? Then you have to know the address of the GPIO and the Value you have to manipulate. It would be nice to have a Pythonic api to change the clk src.

**Comment by yoann-darche** [comment]
I'm not sure if we can update the clock source directly from the register; it might be possible. However, I think the best approach would be to update the µPython API. The important thing to know is that each PWM channel needs to be associated with one of the three available timers (which handle the frequency and duty cycle of the PWM signal). In the timer configuration, we can choose the clock source, can be set at least RC_FAST_CLK (available in light sleep mode) or APB_CLK (the main clock, which shuts down when the ESP enters light sleep).

**Reply by robert-hh** [reply]
There is the open PR #10854 which aims to make the PWM compatible to other ports. It has still one property which is different to other ports. Maybe that blocked merging. So maybe that should settle first, once the author of this PR maintained it.
About you request: when switching between clocks, the PWM frequency will change as well. ATM, PWM uses the APB_CLK, which is exactly 80 MHz. RC_CLK_FAST is about 8.5 MHz, and being based on a RC oscillator I expect it to have a substantial drift with temperature and supply voltage. So it cannot be used for applications requiring a precise and stable frequency.

**Reply by yoann-darche** [reply]
Yes you are certainly right about the drift taht can may occur on the RC_CLK_FAST. 
But in these case the objective is not to have a reliable PWM signal, but have PWM signal when the ESP is in the light sleep mode (ie to control a basic fan, or led).
In the actual version of µPython API, when we setup a GPIO as PWM signal, we can not set the timer properties underlying the PWM. So when we setup a PWM output we don't know on witch inetrnal chanel or timer. It might be an issue when we need to have precise configuration.

Can we just expose the PWM timer thru an API like PWM_TIME(id) then be able to get/set main attribute thru a controlled approach for example disable the possibility of changing the clock source if the timer is already in use in one of the PWM channel

Ideal exemple:
```py
from machine import Pin, Timer, PWM, lightsleep
import time


# PWM
PWM_1 = const(4)
PWM_2 = const(13)

O_PWM1 = None
O_PWM2 = None



def init_hardware():
    
    global O_PWM1, O_PWM2
	
    # Set PWM Timer	
    TM1 = PWM_TIMER(id=0, src_clk = PWM_TIMER_RC_FAST_CLK)	

    # here we select a specific timer	
    O_PWM1 = PWM(Pin(PWM_1,  Pin.OUT), freq=25000, duty_u16=65535, source_timer = TM1) 
    # here we go to the actual process to find the best timer
    O_PWM2 = PWM(Pin(PWM_2,  Pin.OUT), freq=12500, duty_u16=65535)

# Main
def Main():
    
    global O_PWM1, O_PWM2
    
    init_hardware()

    V = 50

    while True:
        
        if V == 50:
            V = 25
        else:
            V = 50
            
        # Set Fan PWM:       
        duty_pct = (100-V) * 65535 // 100    
    
        O_PWM1.duty_u16(duty_pct)
        O_PWM2.duty_u16(duty_pct)
        print('OPWM1',O_PWM1)
        print('OPWM2',O_PWM2)
        
        
        time.sleep(10)
        
        print("[i] Activation du someil légé (20s) ! zZz ")
        lightsleep(20*1000)
        print("[i] WakeUp !")

# Global Start
Main()
```

We may need to add some functions like release() or deinit() to the PWM_TIMER api in order to release the configuration

**Reply by yoann-darche** [reply]
One more thing the changing the clock source impact all timer in case of ESP32-S3, but not on  ESP32-S2, I've reused table form https://github.com/micropython/micropython/pull/10854  and _Extract from ESP32, ESP32-S2 and ESP32-S3 hardware documentation_: 
```
=======================================================  ========  ========  ========  ========  ========
Hardware specification                                      ESP32  ESP32-S2  ESP32-S3  ESP32-C3  ESP32-C6
-------------------------------------------------------  --------  --------  --------  --------  --------
Timers Clock Source
   APB_CLK (80 Mhz)                                           YES       YES       YES       YES	      YES
   REF_TICK (1 Mhz)                                           YES       YES        NO        NO        NO
   RC_FAST_CLK (~ 8Mhz Light sleep compatible)                YES       YES       YES       YES       YES
   XTAL_CLK (40 Mhz)                                           NO       YES       YES       YES       YES
Number of groups (set of timer and channel)                     2         1         1         1         1
Number of timers (per group)                                    4         4         4         4         4
   clock_source independant timer (per group)                   4         4         1         1         1
Number of channels (per group)                                  8         8         8         6         6
-------------------------------------------------------  --------  --------  --------  --------  --------
Different PWM frequencies = (groups * timers)                   8         4         4         4         4
Total PWM channels (Pins, duties) = (groups * channels)        16         8         8         6         6
=======================================================  ========  ========  ========  ========  ========
```

That means setting the source clock for one timer can change the source clock for all timers in the case of the ESP32-S3, but not for other versions. There are also some differences in the available source clocks.

**Comment by yoann-darche** [comment]
API proposal to set the clock source for PWM timer:

```py
    # Set PWM Timer default clock source_timer
    PWM_TIMERS.setDefualtClockSource(PWM_TIMER_RC_FAST_CLK)
    # PWM_SET_DEFAULT_CLOCK_SOURCE(PWM_TIMER_APB_CLK)
    # PWM_SET_DEFAULT_CLOCK_SOURCE(PWM_TIMER_REF_TICK)
    # PWM_SET_DEFAULT_CLOCK_SOURCE(PWM_TIMER_XTAL_CLK)
    
    # Set PWM Timer	
    TM1 = PWM_TIMERS.getTimer(id=0, group=0)	
    TM1.setClockSource(PWM_TIMER_APB_CLK) # will be ignored or rise an error when used with ESP32-S3
    
    # here we select a specific timer
    O_PWM1 = PWM(Pin(PWM_1,  Pin.OUT), freq=25000, duty_u16=65535, source_timer = TM1) 
    # here we go to the actual process to find the best timer
    O_PWM2 = PWM(Pin(PWM_2,  Pin.OUT), freq=12500, duty_u16=65535)
```

**Comment by robert-hh** [comment]
The PWM class itself and it's methods are defined as a common interface at extmod/machine_pwm.c, which defines the class and it's methods. Some freedom exist only in the options of the constructor resp. init method. Selecting the clock source would then typically be an keyword option of the PWM constructor and pwm.init() method with the appropriate class constants for APB_CLK and RC_FAST_CLK. There could be more keyword options to select a specific timer/channel pair. See for instance the SAMD port which allows selecting a `device`, which is indeed a timer, or the MIMXRT port which allows selecting more PWM options besides the common ones.

**Comment by mattytrentini** [comment]
It's a feature _very_ specific to the ESP32 and so an alternative would be to add it to the [esp32 module](https://docs.micropython.org/en/latest/library/esp32.html). Something like:

```py
import esp32
import machine

pwm = machine.PWM(Pin(1))
esp32.pwm_clock_source(pwm, esp32.RC_FAST_CLK)
```

**Reply by yoann-darche** [reply]
I think it is not a good idea, as we need to setup the timer before enabling the PWM. 
Also we may use this approach to set up the default clock source. However, I think that this information will be better situated in the PWM source code.

**Comment by yoann-darche** [comment]
New API Proposal:
```py
# Default case
# The first available group/timer/channel is find automaticly
PWM(Pin(PWM_1,  Pin.OUT), freq=25000, duty_u16=65535)

# Selecting a specific timer assuming that the group is the first one in case of ESP32, in case of ESP32-S2 and ESP31-S3 the group will be ignored
# The clock source will get the previous configuration set on this timer or the default one APB_CLK
# the channel will be the first available
PWM(Pin(PWM_1,  Pin.OUT), freq=25000, duty_u16=65535, timer=0)

# Selecting a specific timer, and setting a clock source
# if ESP32 : group=0, timer=0 and the source clock is set to RC_FAST_CLK
# if ESP32-S2 : group=0 timer=0 and the source clock is set to RC_FAST_CLK
# if ESP32-S3 : group=0 timer=0 and the source clock for all timer will be set to RC_FAST_CLK
# the channel will be the first available
PWM(Pin(PWM_1,  Pin.OUT), freq=25000, duty_u16=65535, timer=0, clock=PWM_RC_FAST_CLK)

# for ESP32 we can add the group information
PWM(Pin(PWM_1,  Pin.OUT), freq=25000, duty_u16=65535, timer=0, group=1, clock=PWM_RC_FAST_CLK)
```

Open question: 
* Do we permit to update of a Timer already in use ?
* if we change a clock source (so a base frequence) do we regenerate all active PWM instance ?

**Reply by robert-hh** [reply]
Answers:

a) Yes. re-configuration of a PWM should always be possible
b) Not automatically. Re-configuration has to be done by the respective Python code.

Besides, to get that feature, someone has to implement it, test it at all ESP32 variants and make a Pull Request. Then it may get into the mainline code. The MicroPython maintainer(s) will most likely not do that themselves and leave it to the community.

**Reply by yoann-darche** [reply]
Ok, I understand that point, I'm interrested by updating the code for these features and I've ESP32, ESP32-S2 and ESP32-S3 to test. 
I'm pretty new in participating to big open source project. I usually code on my own. I can handle the task. But I'm not familliar with the git process (Pull Request/Merge) and so on.
I've mostly read the literature about how to participate to micropython. But i don't really know how to setup the dev environnement (Win11). Do you have any good readme/doc/discussion on how to setup and compile the micorpython for ESP32, it will help me ?

**Reply by robert-hh** [reply]
The best way to develop for MicroPython with Windows is to use Linux, either by using WSL or a Virtual Machine like Virtualbox. 
The setup is sufficiently explained in the Readme for Micripython and the respective port directory. The esp-idf can be a little bit confusing, but it's doable.

**Comment by yoann-darche** [comment]
Ok, I've successfully installed the development environment to recompile MicroPython and successfully generated ports for the ESP32-S2 and ESP32-S3. I've made some basic changes in the machine_pwm.c file to fix the clock source to LEDC_USE_RC_FAST_CLK instead of LEDC_USE_APB_CLK.

And, tada, it works as expected on both targets: the PWM signal continues to run in light sleep mode.

Now, I will start coding a nice API interface to support the clock source.

Based on my tests, I think we could simplify the previous API proposal. In fact, it’s not necessary to specify the timer instance. We might prefer to keep the current algorithm that selects the best timer source based on the expected frequency compared to the existing timer frequency. We would just add 2 new items to compare: _clock_ and _lightSleepEnable_.
```py
# Default case
# The first available group/timer/channel is find automaticly
PWM(Pin(PWM_1,  Pin.OUT), freq=25000, duty_u16=65535)


# Setting a specific clock source
# It will find the first time with a frequence=1K and clock_source = LEDC_USE_RC_FAST_CLK, if not it will select a new timer
# if ESP32 or ESP32-S2 : each timer can have independant clock source
# if ESP32-S3 : the source clock for all timer will be set to LEDC_USE_RC_FAST_CLK
# the channel will be the first available
PWM(Pin(PWM_1,  Pin.OUT), freq=1000, duty_u16=65535, clock=PWM_RC_FAST_CLK)

# and if plan to use the PWM signal in light_sleep mode we will need to add ligthSleepEnable=True
# this adds the GPIO in a special list to "not be disabled" by the light sleep mode
PWM(Pin(PWM_1,  Pin.OUT), freq=1000, duty_u16=65535, clock=PWM_RC_FAST_CLK, ligthSleepEnable=True)
```

**Reply by IhorNehrutsa** [reply]
A small note about the `PWM(Pin(PWM_1,  Pin.OUT),...`
Pin.OUT is not needed and maybe harmful sometimes. 
https://github.com/micropython/micropython/blob/ee8d8b3448f26abafbfe1ed7f8970e02f4f8d232/docs/esp32/tutorial/pwm.rst?plain=1#L104-L115

**Reply by yoann-darche** [reply]
Yes you are rigth, the Pin.OUT is optional and we don't need to specify it, so it can be remove
